core/unix/coredump.c

```diff
@@ -320,41 +329,17 @@ os_dump_core_internal(void)
 bool
 os_dump_core_live(void)
 {
-    static thread_id_t current_dumping_thread_id VAR_IN_SECTION(NEVER_PROTECTED_SECTION) =
-        0;
-    thread_id_t current_id = d_r_get_thread_id();
-#ifdef DEADLOCK_AVOIDANCE
-    dcontext_t *dcontext = get_thread_private_dcontext();
-    thread_locks_t *old_thread_owned_locks = NULL;
-#endif
-
-    if (current_id == current_dumping_thread_id) {
-        return false; /* avoid infinite loop */
-    }
-
-#ifdef DEADLOCK_AVOIDANCE
-    /* First turn off deadlock avoidance for this thread (needed for live dump
-     * to try to grab all_threads and thread_initexit locks).
-     */
-    if (dcontext != NULL) {
-        old_thread_owned_locks = dcontext->thread_owned_locks;
-        dcontext->thread_owned_locks = NULL;
+    uint num_threads;
+    void **drcontexts = NULL;
+    // Suspend all threads including native threads to ensure the memory regions
+    // do not change in the middle of the core dump.
+    if (!dr_suspend_all_other_threads_ex(&drcontexts, &num_threads, NULL,
```

Comment:
Hmm, DR-internal uses normally use synch_with_all_threads directly (to avoid drcontext types being hidden behind void* and to give more control over the flags).  Though for the use case here maybe it's fine. Are we sure this point satisfies the flags passed?  I have a separate comment on ensuring the caller satisfies the requirements for the client-exposed flags.  I think if we update those API docs then it's ok.

Reply:
Done. I've updated api/docs/release.dox to add the reference to the same restrictions as dr_suspend_all_other_threads_ex().